tRPC - Case Study (#18)
NoroffFEU/development-platforms: trpc-case-study.md

```tsx
      <form
          className="ml-5 rounded bg-white p-5"
          onSubmit={(event) => {
            event.preventDefault();
            newNote.mutate({
              title: data.title,
              description: data.description,
            });
            setData({
              title: '',
              description: '',
            });
          }}
        >
```

The repository is configured with a public MongoDB cluster with our prisma model. On submit, the form should now add a new entry to the database with the content of ``title.value`` and ``description.value``. 

![alt-image](tRPC-media/form-example.png)

A very basic layout render of the form as well as alsready existing notes in the database on the right hand side. Submitting the form should now update the database with a new entry:

![alt-image](tRPC-media/database-result.png)

## Strengths

#### tRPC is lightweight
Since tRPC is a protocol or method for accessing backend functionality, it allows for better performing applications without any code generation or runtime bloat.

> Since tRPC uses Typescript, which is a statically typed language, type definitions are checked during compiling rather than runtime, increasing speed of application.

Author: tRPC advertise the product with no build step, but I understand why this part conflicts with the previous section. As I can't properly explain it, I've decide to cut it out from the case-study.

#### tRPC uses Typescript
This can be viewed as both a strenght and a weakness. On one hand, it allows for a uniform coding language and infrastructure between backend and frontend, improving worklflow and communcation between the team. 

> While on the other hand, if Typescript is not the preferred programming language, or if team structure is heavily silo'd between front- and backend tRPC might not be the best option.

Reviewer: ```suggestion
While on the other hand, if Typescript is not the preferred programming language, or if team structure is heavily siloed between front and backend tRPC might not be the best option.
```

#### tRPC works great in a monorepo
With how tRPC communicates between client and server code, you are encouraged to use a singluar repository, and several highlighted features are only accessible if you do so. On a smaller scale, there are many benefits in doing so. It's easier to navigate between client/server code, making changes and ensure typesafety within the entirety of your code base. No more ``uncaught type-errors``.

#### Quick development
- Fastest and easiest way to develop an API. 
- Uses type inferference, which means the type of data is automatically detected, which particularly boosts development in the early stages.
- Integrate with IDE's.

With some familiary to tRPC and well synergized techonologies, such as Next.js and Prisma, you can have functional endpoints up and running within minutes using a pre-configured project package. Such as [T3-app](https://create.t3.gg/).

## Weaknesses

#### tRPC uses Typescript

> While your front-end certainly has knowledge of Javascript, and possibly also well versed in Typescript. The same might not be the case for your backend team, or it might simply not be the preferred techonology. There is a greater variety in programming lanugages typically used for backend than there is for frontend. 

Reviewer: ```suggestion
While a front-end team certainly has knowledge of Javascript and may be well versed in Typescript, the same might not be the case for a backend team. There is a greater variety in programming languages typically used for backend than for frontend, meaning that Typescript may not always be the most suitable choice.
```

> While strict static typing leaves a lot of benefits, it can be frustrating to and cumbersome to learn, especially for junior level developers.

Reviewer: ```suggestion
While strict static typing leaves a lot of benefits, it can be frustrating and cumbersome to learn, especially for junior-level developers.
```

#### Works best in a monorepo

> Since many of tRPC's features benefit from using a singular repository, it might not be the best option if the development team has a need or requirement for heavily silo'd frontend and backend.

Reviewer: ```suggestion
Since many of tRPC's features benefit from using a singular repository, it might not be the best option if the development team has a need or requirement for a heavily siloed frontend and backend.
```

Reviewer: Contractions are quite rare in English and it's typically safer to go with `[-ed]` unless it looks really strange. Siloed looks pretty weird but is indeed correct.

#### Scalability
- In larger scale projects, in a larger team, its often beneficial to come up with a clear set of rules and a schema that everyone can work from. tRPC does not offer this.

> - If your app is getting increasingly more complex and require more nuanced requirements, tRPC might not be the best option.

Reviewer: Why not?

Author: I've resolved this by explaining that other options provide better solutions for increasingly scaling complextion. So while there is nothing wrong with using tRPC in this scenario. The project might benefit more from a different solution.

Similar to ``tRPC``, ``GraphQL`` offers typesafety. ``GraphQL`` is organized in terms of types and fields, instead of endpoints. 

#### Scalability and maintainance
With ``GraphQL`` you can easily add new fields and types to the API without impacting or breaking existing queries. As a result the server code base becomes easier to maintain and allows for API growth with less hassle and complication.

GraphQL is being used by industry juggernaughts such as ``Facebook``, ``Github``, ``Airbnb``, ``Twitter`` and many others.

### Which is better? tRPC or GraphQL?
``tRPC`` is still very much in its infancy, and while many of its features are excellent it cannnot match ``GraphQL's`` scalability and network efficiency. Nor does it need or try to. In my view, they can coexist not as competing technologies, but technologies that both offer their strength and weaknesses depending on use cases.

*Use GraphQL if:*
- Its a large scale project, with the need of well defined schemas, maintenance and future development requirements.

> - if your team is either heavily silo'd or use multiple programming languages.

Reviewer: ```suggestion
- if your team is either heavily siloed or use multiple programming languages.
```

*Use tRPC if:*

> - You are looking to learn to create your own API's.

Reviewer: ```suggestion
- You are looking to learn to create your own APIs.
```
In this case `API's` could mean:
- Belonging to API (the API's test methods are passing)
- API is ... (the API's broken again)

But not:
- More than one API (the APIs are working correctly)

- You are working in a smaller development team or work as a freelancer.
- You use Typescript as your preferred programming language.
- You are developing a smaller scope application.

## Summary

> tRPC offers an excellent introduction to creating API's. Using it is intuitive, fast and efficient. With some experience, you can have functional endpoints up and running within minutes. Whether you are a junior when it comes to backend or web-development in general, or an experienced full-stack developer tRPC should offer a lot of welcome features for your next project - increasing your workflow, maintaining your code base and overall helping you become an overall more productive developer. 

Reviewer: ```suggestion
tRPC offers an excellent introduction to creating APIs. Using it is intuitive, fast and efficient. With some experience, you can get functional endpoints up and running within minutes. Whether you are a junior when it comes to backend or web development in general, or an experienced full-stack developer tRPC should offer a lot of welcome features for your next project - increasing your workflow, maintaining your code base and overall helping you become an overall more productive developer. 
```
